From 236a02635efa2af95eba1a39d05e6ebb60926cb8 Mon Sep 17 00:00:00 2001
From: Andrii Sultanov <andriy.sultanov@vates.tech>
Date: Wed, 26 Nov 2025 10:47:27 +0000
Subject: [PATCH] vhd_tool_wrapper: Add parse_header to determine allocated
 blocks

Vhd_tool_wrapper.run_vhd_tool is hardcoded to read the progress percentage
printed by vhd-tool, so use Qcow_tool_wrapper.run_qcow_tool to run vhd-tool
since it's generic enough to allow that.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>
---
 ocaml/xapi/qcow_tool_wrapper.ml  | 32 ++++++++++++++++++++++++++++++++
 ocaml/xapi/qcow_tool_wrapper.mli |  4 ++++
 ocaml/xapi/vhd_tool_wrapper.ml   |  2 ++
 3 files changed, 38 insertions(+)

diff --git a/ocaml/xapi/qcow_tool_wrapper.ml b/ocaml/xapi/qcow_tool_wrapper.ml
index b894ff05d..849ea3d4c 100644
--- a/ocaml/xapi/qcow_tool_wrapper.ml
+++ b/ocaml/xapi/qcow_tool_wrapper.ml
@@ -62,6 +62,38 @@ let read_header qcow_path =
     (fun () -> Unix.close pipe_writer) ;
   pipe_reader
 
+let read_vhd_header path =
+  let vhd_tool = !Xapi_globs.vhd_tool in
+  let args = ["read_headers"; path] in
+  let pipe_reader, pipe_writer = Unix.pipe ~cloexec:true () in
+
+  let progress_cb _ = () in
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () -> run_qcow_tool vhd_tool progress_cb args ~output_fd:pipe_writer)
+    (fun () -> Unix.close pipe_writer) ;
+  pipe_reader
+
+let parse_header pipe_reader =
+  let ic = Unix.in_channel_of_descr pipe_reader in
+  let buf = Buffer.create 4096 in
+  let json = Yojson.Basic.from_channel ~buf ~fname:"header.json" ic in
+  In_channel.close ic ;
+  let cluster_size =
+    1 lsl Yojson.Basic.Util.(member "cluster_bits" json |> to_int)
+  in
+  let cluster_list =
+    Yojson.Basic.Util.(member "data_clusters" json |> to_list |> List.map to_int)
+  in
+  (cluster_size, cluster_list)
+
+let parse_qcow_header qcow_path =
+  let pipe_reader = read_header qcow_path in
+  parse_header pipe_reader
+
+let parse_vhd_header qcow_path =
+  let pipe_reader = read_vhd_header qcow_path in
+  parse_header pipe_reader
+
 let send ?relative_to (progress_cb : int -> unit) (unix_fd : Unix.file_descr)
     (path : string) (_size : Int64.t) =
   let qcow_of_device =
diff --git a/ocaml/xapi/qcow_tool_wrapper.mli b/ocaml/xapi/qcow_tool_wrapper.mli
index 51c3c6265..8c41b938d 100644
--- a/ocaml/xapi/qcow_tool_wrapper.mli
+++ b/ocaml/xapi/qcow_tool_wrapper.mli
@@ -23,3 +23,7 @@ val send :
   -> string
   -> int64
   -> unit
+
+val parse_vhd_header : string -> int * int list
+
+val parse_qcow_header : string -> int * int list
diff --git a/ocaml/xapi/vhd_tool_wrapper.ml b/ocaml/xapi/vhd_tool_wrapper.ml
index 2a3d40ae5..de9cf0590 100644
--- a/ocaml/xapi/vhd_tool_wrapper.ml
+++ b/ocaml/xapi/vhd_tool_wrapper.ml
@@ -112,6 +112,8 @@ let receive progress_cb format protocol (s : Unix.file_descr)
   in
   run_vhd_tool progress_cb args s s' path
 
+let parse_header = Qcow_tool_wrapper.parse_vhd_header
+
 let send progress_cb ?relative_to (protocol : string) (dest_format : string)
     (s : Unix.file_descr) (path : string) (size : Int64.t) (prefix : string) =
   let vhd_of_device =
